src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,862 @@
+                CreateDisplayText(expression, syntaxFacts, FeaturesResources.Introduce_parameter_for_0), action.Value.Item1, isInlinable: false), textSpan);
+            context.RegisterRefactoring(new CodeActionWithNestedActions(
+                CreateDisplayText(expression, syntaxFacts, FeaturesResources.Introduce_parameter_for_all_occurrences_of_0), action.Value.Item2, isInlinable: false), textSpan);
+        }
+
+        /// <summary>
+        /// Introduces a new parameter and refactors all the call sites.
+        /// </summary>
+        public async Task<Solution> IntroduceParameterAsync(Document document, TExpressionSyntax expression,
+            IMethodSymbol methodSymbol, SyntaxNode containingMethod, bool allOccurrences, bool trampoline, bool overload,
+            CancellationToken cancellationToken)
+        {
+            var parameterName = await GetNewParameterNameAsync(document, expression, cancellationToken).ConfigureAwait(false);
+
+            var methodCallSites = await FindCallSitesAsync(document, methodSymbol, cancellationToken).ConfigureAwait(false);
+
+            return await RewriteSolutionAsync(document,
+                expression, methodSymbol, containingMethod, allOccurrences, parameterName, methodCallSites,
+                trampoline, overload, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Gets the parameter name, if the expression's grandparent is a variable declarator then it just gets the
+        /// local declarations name. Otherwise, it generates a name based on the context of the expression.
+        /// </summary>
+        private static async Task<string> GetNewParameterNameAsync(Document document, TExpressionSyntax expression, CancellationToken cancellationToken)
+        {
+            if (ShouldRemoveVariableDeclaratorContainingExpression(document, expression, out var varDeclName, out _))
+            {
+                return varDeclName;
+            }
+
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var semanticFacts = document.GetRequiredLanguageService<ISemanticFactsService>();
+            return semanticFacts.GenerateNameForExpression(semanticModel, expression, capitalize: false, cancellationToken);
+        }
+
+        /// <summary>
+        /// Determines if the expression's grandparent is a variable declarator and if so,
+        /// returns the name
+        /// </summary>
+        protected static bool ShouldRemoveVariableDeclaratorContainingExpression(
+            Document document, TExpressionSyntax expression, [NotNullWhen(true)] out string? varDeclName, [NotNullWhen(true)] out SyntaxNode? localDeclaration)
+        {
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            localDeclaration = expression.FirstAncestorOrSelf<SyntaxNode>(node => syntaxFacts.IsLocalDeclarationStatement(node));
```

Comment:
this is not hte right way to do it.  you need to validate the varidecl's parent is a local declaration.  For example, you would return true in this case:

```
var v = () =>
{
    for (var y = a * b; ;) { }
};
```

That's because 'var y' is a variable declarator *and* there is an ancestor local decl statement.  Need test for thsi case :)

Reply:
fixed, will add tests